### PR TITLE
Core/UnitAI: Rework creature-controlled player behavior.

### LINKED ISSUE #4723: [Bug]Domination makes the player "stick to one place"
Bug:
Whenever a mob successfully casts Domination on a player, the character "gets stuck in one place" while still being able to move, but whenever the player uses a abilities like Prowl or Bear Form he comes back to the same place(the video will explain a lot more). I've tested this only on Harbinger Skyriss and Dark Rune Controller, but i think this accounts for all spells of the same name and perhaps all "mind control" spells as well.

Info:
Domination - http://old.wowhead.com/spell=51503
Video of the bug - http://www.dailymotion.com/video/xnj0dp_bug-report-domination_videogames

Core revision: Trinity Core 991934911edb+
Database revision: TDB 335.11.45
Addons: Anticheat1

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/65491-bug-domination-makes-the-player-stick-to-one-place?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Treeston**
Core/UnitAI: Rework creature-controlled player behavior.
- Removed hacked control mechanism, use proper PlayerAI instead
- Port old hacky code to new SimpleCharmedPlayerAI class
- Make adjustments to aforementioned code to fix bugs:
  - Properly clean up movement after charm ends
  - Only try to attack a target if charmer is engaged in combat
- Allow CreatureAI to override the AI applied to players charmed by its creature on a case by case basis

Fixes #4723.

### CODE REVIEW

**Comment by P-Kito** [comment]
This is for creatures that charm players right?

Charmed players blow their cooldown, this should also be implemented in this commit.

**Comment by Treeston** [comment]
For now I want to merge a basic generic player AI that works. After that, we can iterate easily on this to create smarter charmed player behavior.

**Comment by Keader** [comment]
This should be fixed on future? #15379

**Comment by Treeston** [comment]
Yes, that's what I'm referring to.

File: src/server/game/Entities/Player/Player.h [hunk 1]
```diff
@@ -1022,6 +1023,8 @@ class Player : public Unit, public GridObject<Player>
         explicit Player(WorldSession* session);
         ~Player();
 
+        PlayerAI* AI() const { return (PlayerAI*)i_AI; }
```
Comment (DDuarte) [inline src/server/game/Entities/Player/Player.h]: Avoid C style casts https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Res-casts-named

**Comment by Treeston** [comment]
Change summary:
- Moved all player AI related classes from UnitAI code to its own PlayerAI.cpp/PlayerAI.h.
- Ranged classes now use ranged attacks instead of closing to melee range. This excludes healers (anecdotally blizzlike).

**Comment by P-Kito** [comment]
Looks good, well done

**Comment by P-Kito** [comment]
Spells! We need Spells! :dart:

**Comment by Treeston** [comment]
Sure, once we figure out #16687 so it looks less stupid when I make them cast.

File: src/server/game/AI/PlayerAI/PlayerAI.cpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            case ITEM_SUBCLASS_WEAPON_THROWN:
+                rangedAttackSpell = SPELL_THROW;
+                break;
+            case ITEM_SUBCLASS_WEAPON_WAND:
+                rangedAttackSpell = SPELL_SHOOT_WAND;
+                break;
+        }
+    }
+
+    std::cout << "Selected " << rangedAttackSpell << std::endl;
```
Comment (robinsch) [inline src/server/game/AI/PlayerAI/PlayerAI.cpp]: .

Comment (Treeston) [inline src/server/game/AI/PlayerAI/PlayerAI.cpp]: Already fixed in 1fbe7979fff0e67c6d390113f8e4a1099db8193d

**Pull request merged**